metricbeat/helper/server/http/http_test.go

```diff
@@ -22,17 +22,86 @@ package http
 import (
 	"bytes"
 	"context"
+	"crypto/tls"
 	"fmt"
+	"io/ioutil"
+	"net"
 	"net/http"
+	"strconv"
 	"testing"
-	"time"
-
-	"github.com/elastic/beats/metricbeat/helper/server"
 
 	"github.com/stretchr/testify/assert"
+
+	"github.com/elastic/beats/metricbeat/helper/server"
 )
 
-func GetHttpServer(host string, port int) (server.Server, error) {
+var flagtests = []struct {
+	testName, inputMessage, connectionType, connectionMethod, expectedOutput string
+	expectedHTTPCode                                                         int
+}{
+	{"HTTP GET", `"@timestamp":"2016-05-23T08:05:34.853Z"`, "HTTP", "GET", "HTTP Server accepts data via POST", 200},
+	{"HTTPS GET", `"@timestamp":"2016-05-23T08:05:34.853Z"`, "HTTPS", "GET", "HTTPS Server accepts data via POST", 200},
+	{"HTTP POST", `"@timestamp":"2016-05-23T08:05:34.853Z"`, "HTTP", "POST", `"@timestamp":"2016-05-23T08:05:34.853Z"`, 202},
+	{"HTTPS POST", `"@timestamp":"2016-05-23T08:05:34.853Z"`, "HTTPS", "POST", `"@timestamp":"2016-05-23T08:05:34.853Z"`, 202},
+}
+
+func TestHTTPServers(t *testing.T) {
+	for _, test := range flagtests {
+		t.Run(test.testName, func(t *testing.T) {
+			host := "127.0.0.1"
+			port := 40050
+			svc, err := getHTTPServer(t, host, port, test.connectionType)
+			if err != nil {
+				t.Error(err)
+				t.FailNow()
+			}
+
+			svc.Start()
+			defer svc.Stop()
+			// make sure server is up before writing data into it.
+			var conn net.Conn
+			for {
```

Comment:
this is cool, but let's avoid some (very improbable) infinite loops.

What about adding a function that adds a timeout, kindof (feel free to dismiss and add your own):
```
const (
	checkServerReadyTimeout = 5 * time.Second
	checkServerReadyTick    = 100 * time.Millisecond
)

func checkServerReady(host string, port int) error {

	var conn net.Conn
	var err error

	ctx, cancel := context.WithTimeout(context.TODO(), checkServerReadyTimeout)
	defer cancel()
	ticker := time.NewTicker(checkServerReadyTick)
	defer ticker.Stop()
	for {
		select {
		case <-ticker.C:
			conn, err = net.Dial("tcp", net.JoinHostPort(host, strconv.Itoa(int(port))))
			if conn != nil {
				_ = conn.Close()
			}
			if err != nil {
				break
			}
			return nil
		case <-ctx.Done():
			return fmt.Errorf("HTTP server at %s:%d never responded: %+v", host, port, err)
		}
	}

}
```

Reply:
> :) I have made some more changes trying to understand why Travis ci is failing.. Locally the tests happen OK!

same here
will dive into this later

Reply:
Go is very powerful... I've got used to work on .c proprietary code without string.h lol.... this is way better. Nice context library!

Reply:
Moving the variables inside the function. 
I have changed a little the return and the break in this new for/select pattern!